### PR TITLE
[WIP] Add support for 2.10

### PR DESCRIPTION
**Pull request by nrinaudo**
This (mostly) adds support for 2.10. Only one compilation problem remains:

```scala
error] /contribs/contextual/core/src/main/scala/contextual/contextual.scala:51: macro implementation has incompatible shape:
[error]  required: (c: scala.reflect.macros.Context)(exprs: c.Expr[contextual.Interpolator.Embedded[Prefix.this.interpolator.Inputs,Prefix.this.interpolator.type]]*): c.Expr[Any]
[error]  found   : (c: contextual.compat.Context)(exprs: c.Tree*): c.Tree
[error] type mismatch for parameter exprs: c.Expr[contextual.Interpolator.Embedded[Prefix.this.interpolator.Inputs,Prefix.this.interpolator.type]]* does not conform to c.universe.Tree*
[error]       macro Macros.contextual[C, P]
```

I'm making a PR as is following our [gitter conversation](https://gitter.im/propensive/contextual?at=5851cfcafb22792b3b64e9c6).

### CODE REVIEW

**Comment by nrinaudo** [comment]
[This](https://github.com/nrinaudo/contextual/blob/ddaa43c1ed05cd2f6dfb1befa599346bc4f5c690/core/src/main/scala/contextual/contextual.scala#L117)  still fails to compile - I don't know how to get a quasiquote to return anything but a `c.Tree`. Do I need to go through `reify`, or is there something clever to do there?

File: core/src/main/scala/contextual/contextual.scala [hunk 1]
```diff
@@ -91,7 +91,7 @@ trait Interpolator extends Interpolator.Parts { interpolator =>
   /** The macro evaluator that defines what code will be generated for this `Interpolator`. The
     * default implementation constructs a new runtime `Contextual` object, and invokes the
     * `evaluate` method on the `Interpolator`. */
-  def evaluator(contexts: Seq[Ctx], contextual: Contextual[StaticPart]): contextual.context.Tree = {
+  def evaluator(contexts: Seq[Ctx], contextual: Contextual[StaticPart]): contextual.context.Expr[Any] = {
```
Comment (propensive) [inline core/src/main/scala/contextual/contextual.scala:94]: I think we should leave the return type as `context.Tree`. You'll need to implement a method, say, `compat.macroReturn`, which always returns `context.Tree`, but takes `context.Expr` in the Scala 2.10 implementation, and `context.Tree` in the 2.11+ version.

File: core/src/main/scala/contextual/macros.scala [hunk 2]
```diff
@@ -19,8 +19,7 @@ import language.experimental.macros
 /** Object containing the main macro providing Contextual's functionality. */
 object Macros {
   def contextual[C <: Context, I <: Interpolator { type Ctx = C }: c.WeakTypeTag]
-  (c: compat.Context)(exprs: c.Tree*): c.Tree = {
-  //(c: foobar.Context)(exprs: c.Tree*): c.Tree = {
+  (c: compat.Context)(exprs: c.Expr[Any]*): c.Expr[Any] = {
```
Comment (propensive) [inline core/src/main/scala/contextual/macros.scala:22]: Let's leave the `exprs` type as `c.Tree`, and use a new method in `compat` to deal with it appropriately in the different Scala versions.

File: core/src/main/scala/contextual/contextual.scala [hunk 3]
```diff
@@ -67,7 +67,7 @@ trait Interpolator extends Interpolator.Parts { interpolator =>
     val context: compat.Context = null
 
     /** The expressions that are substituted into the interpolated string. */
-    def expressions: Seq[context.Tree] = Nil
+    def expressions: Seq[context.Expr[Any]] = Nil
```
Comment (propensive) [inline core/src/main/scala/contextual/contextual.scala:70]: Leave this as `context.Tree`, and call `.map(compat.toTree)` on the expressions value where necessary. You'll need to implement `compat.toTree` to call `.tree` on an `c.Expr` in 2.10, or just be the identity in 2.11/2.12.

**Review by propensive — CHANGES_REQUESTED** [review]

File: core/src/main/scala/contextual/contextual.scala (hunk 1 above)
Comment (nrinaudo) [reply core/src/main/scala/contextual/contextual.scala:94]: I must admit I've no idea how to turn a `c.Expr`  into a `c.Tree`, aside from using [macro-compat](https://github.com/milessabin/macro-compat)...

**Comment by nrinaudo** [comment]
Not much point now that macro-compat is used.

**Pull request closed without merging**